created how to set up a hardhat project for rsk article (#787)
rsksmart/devportal: kb/hardhat-project-on-rsk.md

### 2.3. Hardhat Network

Hardhat comes built-in with a special network called hardhat. When using this network, an instance of the [Hardhat Network](https://hardhat.org/hardhat-network/) will be automatically created when you run a task, script or test your smart contracts.

Hardhat Network has first-class support of Solidity. It always knows which smart contracts are being run and exactly what they do and why they fail. Learn more about it [here](https://hardhat.org/hardhat-network/).

### 2.4. JSON-RPC based networks

These are networks that connect to an external node. Nodes can run on your computer, like Ganache, or remotely, like RSK Testnet. We'll be configuring a network connection named `rsktestnet` for this tutorial.

Include the following code in your `hardhat.config.js` file

```
/**
 * @type import('hardhat/config').HardhatUserConfig
 */
module.exports = {
  solidity: '0.7.3',
  networks: {
    rsktestnet: {
      chainId: 31,
      url: 'https://public-node.testnet.rsk.co/',
      gasPrice: Math.floor(minimumGasPriceTestnet * TESTNET_GAS_MULT),
      gasMultiplier: TESTNET_GAS_MULT,
      accounts: {
        mnemonic: mnemonic,
        initialIndex: 0,

>         path: "m/44'/60'/0'/0",

Reviewer: Does it make sense to mention [RSK Testnet checksum](https://developers.rsk.co/rsk/architecture/account-based) path instead of default ETH?

`m/44'/37310'/0'/0"`

Reviewer 2: @Ginowine

`npm install --save-dev @nomiclabs/hardhat-ethers ethers @nomiclabs/hardhat-waffle ethereum-waffle chai`

### 4. Writing and compiling smart contracts

We're going to create a simple smart contract that implements a token that can be transferred. Token contracts are most frequently used to exchange or store value. Find below some logic we implemented by the contract.

* There is a fixed total supply of tokens that can't be changed.
* The entire supply is assigned to the address that deploys the contract.
* Anyone can receive tokens.
* Anyone with at least one token can transfer tokens.
* Get Token balance of a given account

### 4.1 Creating the  smart contracts

Start by creating a new directory called `contracts` and create a file inside the directory called `Token.sol`.

```
mkdir contracts
touch contracts/Token.sol
code contracts/Token.sol
```

Copy the below initial set up code for the Token transfer contract into your integrated development environment. We’ll implement the functions in the smart contract as we go along.

> Note that there are several places marked with `// TODO` as comments.

> > This tutorial will guide you on how to complete this implementation.

Reviewer 2: ```suggestion
```

* `address payable`: Same as address, but with the additional members transfer and send.

The idea behind this distinction is that address payable is an address you can send RBTC to, while a plain address cannot be sent RBTC. Read more on [Variable Declaration in Solidity](https://docs.soliditylang.org/en/v0.7.1/style-guide.html?highlight=bool#mappings)

Next, we store each account balance in a key/value pair map. The line `mapping(address => uint256) balances;`maps an address field that stores the address of the account, and a uint256 field that stores the balance of Tokens in the mapped address. 

The constructor function gets called when the smart contract is deployed to the blockchain. The constructor would be initialised upon contract creation. Read about [Constructors in Solidity](https://solidity.readthedocs.io/en/v0.7.1/contracts.html#constructor).

### 4.5. Implementing the smart contract functions

### Implementing the constructor 

```
constructor() {
        // The totalSupply is assigned to transaction sender, which is the account
        // that is deploying the contract.
        balances[msg.sender] = totalSupply;
        owner = msg.sender;
    }
```

> The line `balances[msg.sender] = totalSupply;` assigns the address of the person deploying the contract to the balances address field, and then assigns the `totalSupply` to the `uint256` field of the balances variable. The line `owner = msg.sender;` assigns the address of the person deploying the contract to the owner variable, which is of type address. 

Reviewer 2: ```suggestion
The line `balances[msg.sender] = totalSupply;` assigns the address of the person deploying the contract to the balances address field, and then assigns the `totalSupply` to the `uint256` field of the balances variable. The line `owner = msg.sender;` assigns the address of the person deploying the contract to the owner variable, which is of type `address`. 
```

        console.log("Trying to send %s tokens to %s", amount, to);
        // Check if the transaction sender has enough tokens.
        // If `require`'s first argument evaluates to `false` then the
        // transaction will revert.
        require(balances[msg.sender] >= amount, "Not enough tokens");

        // Transfer the amount.
        balances[msg.sender] -= amount;
        balances[to] += amount;
    }
```

The function transfer takes two parameters, the first is the address to which Tokens should be transfered to, and the second parameter is the amount of Tokens to be sent. The external modifier makes this function only callable from outside the contract. The functions requires that balance of sender should be greater than amount to be sent, if this condition is false, then an error message is thrown, saying "not enough Tokens". If the balance amount is greater than amount to be sent, then the amount is deducted from the senders balance and then credited to the receivers account.

### 4.7. Implementing the balanceOf Function

```
    function balanceOf(address account) external view returns (uint256) {
        return balances[account];
    }
```

> The function `balanceOf` takes an address as parameter, and then returns the current balance of that address to the caller of the function. The view keyword in the declaration means that the function will not modify state of contract, while the returns keyword indicate that the function will return a `uint256` value.

Reviewer 2: ```suggestion
The function `balanceOf` takes an address as a parameter and then returns the current balance of that address to the caller of the function. The view keyword in the declaration means that the function will not modify the state of the contract, while the returns keyword indicate that the function will return a `uint256` value.
```

        balances[msg.sender] = totalSupply;
        owner = msg.sender;
    }

    /**
     * A function to transfer tokens.
     *
     * The `external` modifier makes a function *only* callable from outside
     * the contract.
     */
    function transfer(address to, uint256 amount) external {
        console.log("Sender balance is %s tokens", balances[msg.sender]);
        console.log("Trying to send %s tokens to %s", amount, to);
        // Check if the transaction sender has enough tokens.
        // If `require`'s first argument evaluates to `false` then the
        // transaction will revert.
        require(balances[msg.sender] >= amount, "Not enough tokens");

        // Transfer the amount.
        balances[msg.sender] -= amount;
        balances[to] += amount;
    }

    /**
     * Read only function to retrieve the token balance of a given account.
     *
     * The `view` modifier indicates that it doesn't modify the contract's
     * state, which allows us to call it without executing a transaction.
     */
    function balanceOf(address account) external view returns (uint256) {
        return balances[account];
    }
}
```

 

### 4.8. Compiling contracts

> To compile the contract run `npx hardhat compile` in your terminal. The `compile` task is one of the built-in tasks.

Reviewer 2: ```suggestion
To compile the contract, run `npx hardhat compile` in your terminal. The `compile` task is one of the built-in tasks.
```

     // TODO 
    });
  });
});
```

### 5.1. Lets set up our test environment

```
const { expect } = require("chai");

describe("Token contract", function () {
  let Token;
  let hardhatToken;
  let owner;
  let addr1;
  let addr2;
  let addrs;

  beforeEach(async function () {
    Token = await ethers.getContractFactory("Token");
    [owner, addr1, addr2, ...addrs] = await ethers.getSigners();

    hardhatToken = await Token.deploy();
  });
```

*The above code enables proper setting of our test environment*

The line `const { expect } = require("chai");` imports `Chai` which is an assertions library. These asserting functions are called "matchers", and the ones we're using here actually come from Waffle.

This is why we're using the `@nomiclabs/hardhat-waffle` plugin, which makes it easier to assert values from Ethereum.

The next line `describe` is a Mocha function that allows you to organize your tests. It's not actually needed, but having your tests organized makes debugging them easier. All Mocha functions are available in the global scope. `describe` receives the name of a section of your test suite, and a callback.

> The callback must define the tests of that section. This callback can't be an async function. Mocha has four functions that let you hook into the test runner's lifecyle. These are: `before`, `beforeEach`, `after`, `afterEach`.

Reviewer 2: ```suggestion
The callback must define the tests of that section. This callback can't be an async function. Mocha has four functions that let you hook into the test runner's lifecycle. These are: `before`, `beforeEach`, `after`, `afterEach`.
```

> They're very useful to setup the environment for tests, and to clean it up after they run. A common pattern is to declare some variables, and assign them in the`before` and `beforeEach` callbacks.`beforeEach` will run before each test, re-deploying the contract every time. It receives a callback, which can be async.

Reviewer 2: ```suggestion
They're very useful to set up the environment for tests and to clean it up after they run. A common pattern is to declare some variables, and assign them in the `before` and `beforeEach` callbacks.`beforeEach` will run before each test, re-deploying the contract every time. It receives a callback, which can be async.
```

> The lines `Token = await ethers.getContractFactory("Token");` and `[owner, addr1, addr2, ...addrs] = await ethers.getSigners();`,  get's the ContractFactory and Signers.

Reviewer 2: ```suggestion
The lines `Token = await ethers.getContractFactory("Token");` and `[owner, addr1, addr2, ...addrs] = await ethers.getSigners();`,  gets the ContractFactory and Signers.
```

### 5.2. Testing to set the right owner

```
    it("Should set the right owner", async function () {
      expect(await hardhatToken.owner()).to.equal(owner.address);
    });
```
*This test expects the owner variable stored in the contract to be equal to our Signer's owner.*

`it` is another Mocha function. This is the one you use to define your tests, it receives the test name, and a callback function. You might have noticed the use of `async` in it's callback function. Interacting with the Ethereum network and smart contracts are asynchronous operations, hence most APIs and libraries use JavaScript's Promise for returning values. This use of async will allow us to `await` the calls to our contract and the Hardhat Network node.

The next line `expect` receives a value, and wraps it in an assertion object. These objects have a lot of utility methods to assert values.

### 5.3. Testing totalSupply of Token is equal to owner's balance 

```
it("Should assign the total supply of tokens to the owner", async function () {
  const ownerBalance = await hardhatToken.balanceOf(owner.address);
  expect(await hardhatToken.totalSupply()).to.equal(ownerBalance);
});
```

> The above test assigns the total number of Tokens to the owner, and make sures the total number of tokens is equal to the onwer's balance.

Reviewer 2: ```suggestion
The above test assigns the total number of Tokens to the owner and makes sure the total number of tokens is equal to the owner's balance.
```

The line `const addr1Balance = await hardhatToken.balanceOf(addr1.address);` calls the contract's balanceOf function on addr1, and assigns the value to the variable `addr1Balance`. It then expect the `addr1Balance` to be equal to 50. 

The next line `await hardhatToken.connect(addr1).transfer(addr2.address, 50);` uses the `.connect()` function to send a transaction from another account to a second account. The next line then checks the balance of `addr2`, and expects the balance to be equal to 50. 

### 5.5. Testing to make sure sender has enough Tokens

```
 it("Should fail if sender doesn’t have enough tokens", async function () {
      const initialOwnerBalance = await hardhatToken.balanceOf(owner.address);
      await expect( hardhatToken.connect(addr1).transfer(owner.address, 1)
      ).to.be.revertedWith("Not enough tokens");
      expect(await hardhatToken.balanceOf(owner.address)).to.equal(
        initialOwnerBalance
      );
    });
```

The line `const initialOwnerBalance = await hardhatToken.balanceOf(owner.address);`, checks and returns the initial balance of the owner's address. The line `await expect( hardhatToken.connect(addr1).transfer(owner.address, 1)

>       ).to.be.revertedWith("Not enough tokens");`, try's to send 1 token from addr1 (0 tokens) to owner's address (1000000 tokens). 

Reviewer 2: ```suggestion
      ).to.be.revertedWith("Not enough tokens");`, tries to send 1 token from addr1 (0 tokens) to owner's address (1000000 tokens). 
```

      expect(await hardhatToken.balanceOf(owner.address)).to.equal(
        initialOwnerBalance
      );
    });

    it("Should update balances after transfers", async function () {
      const initialOwnerBalance = await hardhatToken.balanceOf(owner.address);

      await hardhatToken.transfer(addr1.address, 100);

      await hardhatToken.transfer(addr2.address, 50);

      const finalOwnerBalance = await hardhatToken.balanceOf(owner.address);
      expect(finalOwnerBalance).to.equal(initialOwnerBalance - 150);

      const addr1Balance = await hardhatToken.balanceOf(addr1.address);
      expect(addr1Balance).to.equal(100);

      const addr2Balance = await hardhatToken.balanceOf(addr2.address);
      expect(addr2Balance).to.equal(50);
    });
  });
});
```

### Running the test

Open a new terminal in the project directory.

Enter the command below into the terminal:

```
npx hardhat test
```
 

### 6. Create smart contract deployment script and dry-run the deployment

Create a new directory called scripts inside our project root directory and create a new file called deploy.js

```
mkdir scripts
touch scripts/deploy.js
code scripts/deploy.js
```

> Past the below code into your deploy.js file

Reviewer 2: ```suggestion
Paste the below code into your deploy.js file
```

```
async function main() {
    const [deployer] = await ethers.getSigners();
  
    console.log("Deploying contracts with the account:", deployer.address);
  
    console.log("Account balance:", (await deployer.getBalance()).toString());
  
    const Token = await ethers.getContractFactory("Token");
    const token = await Token.deploy();
  
    console.log("Token address:", token.address);
  }
  
  main()
    .then(() => process.exit(0))
    .catch((error) => {
      console.error(error);
      process.exit(1);
    });
```

> Run the below command to deploy your application

Reviewer 2: ```suggestion
Run the command below to deploy your application
```
